From 1c4b0ac86a0cf4a47ebb084321b0269a1264ab88 Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Wed, 8 Apr 2026 13:56:46 -0700
Subject: [PATCH] Handle concurrency issues in SDP record creation

If deletion of an SDP record is attempted before the record has been
fully created, add that record to a pending queue and remove it on the
next call of on_create_record_event.

Bug: 415689442
Bug: 413510838
Test: mmm packages/modules/Bluetooth
Flag: EXEMPT BUGFIX
Tag: #security
Merged-In: Idd00b6bfece9fdf04475a9ee310412eccc3240e2
Change-Id: I9d858b3509d246595cc00864d012ac616e33f101
---
 system/btif/src/btif_sdp_server.cc | 52 ++++++++++++++++++++++++++----
 1 file changed, 46 insertions(+), 6 deletions(-)

diff --git a/system/btif/src/btif_sdp_server.cc b/system/btif/src/btif_sdp_server.cc
index 6c073daf61..d8e5202bb3 100644
--- a/system/btif/src/btif_sdp_server.cc
+++ b/system/btif/src/btif_sdp_server.cc
@@ -36,6 +36,7 @@
 #include <string.h>
 
 #include <mutex>
+#include <vector>
 
 #include "bta/include/bta_sdp_api.h"
 #include "bta/sys/bta_sys.h"
@@ -54,14 +55,12 @@ using namespace bluetooth;
 
 // Protects the sdp_slots array from concurrent access.
 static std::recursive_mutex sdp_lock;
+static std::vector<int> sdp_record_creation_pending_ids;
 
-/**
- * The need for a state variable have been reduced to two states.
- * The remaining state control is handled by program flow
- */
 typedef enum {
   SDP_RECORD_FREE = 0,
   SDP_RECORD_ALLOCED,
+  SDP_RECORD_CREATE_INITIATED,
 } sdp_state_t;
 
 typedef struct {
@@ -260,13 +259,15 @@ static const sdp_slot_t* start_create_sdp(int id) {
   }
 
   std::unique_lock<std::recursive_mutex> lock(sdp_lock);
-  if (sdp_slots[id].state != SDP_RECORD_ALLOCED) {
+  if (sdp_slots[id].state == SDP_RECORD_FREE) {
     /* The record have been removed before this event occurred - e.g. deinit */
     log::error("failed - state for id {} is sdp_slots[id].state = {} expected {}", id,
                sdp_slots[id].state, SDP_RECORD_ALLOCED);
     return NULL;
   }
 
+  sdp_slots[id].state = SDP_RECORD_CREATE_INITIATED;
+
   return &(sdp_slots[id]);
 }
 
@@ -301,14 +302,27 @@ bt_status_t remove_sdp_record(int record_id) {
 
   bluetooth_sdp_record* record;
   bluetooth_sdp_types sdp_type = SDP_TYPE_RAW;
+
   {
     std::unique_lock<std::recursive_mutex> lock(sdp_lock);
+
+    if (sdp_slots[record_id].state == SDP_RECORD_CREATE_INITIATED) {
+      sdp_record_creation_pending_ids.push_back(record_id);
+      return BT_STATUS_SUCCESS;
+    }
+    if (sdp_slots[record_id].state == SDP_RECORD_FREE) {
+      log::warn("Sdp Server attempted to remove record already removed or never "
+                         "created");
+      return BT_STATUS_DONE;
+    }
     record = sdp_slots[record_id].record_data;
     if (record != NULL) {
       sdp_type = record->hdr.type;
     }
   }
+
   tBTA_SERVICE_ID service_id = 0;
+
   switch (sdp_type) {
     case SDP_TYPE_MAP_MAS:
       service_id = BTA_MAP_SERVICE_ID;
@@ -340,10 +354,23 @@ bt_status_t remove_sdp_record(int record_id) {
     BTA_SdpRemoveRecordByUser(INT_TO_PTR(handle));
     return BT_STATUS_SUCCESS;
   }
-  log::verbose("Sdp Server - record already removed - or never created");
+
   return BT_STATUS_DONE;
 }
 
+static void sdp_check_pending_records() {
+  std::vector<int> pending_ids;
+  {
+    std::unique_lock<std::recursive_mutex> lock(sdp_lock);
+    pending_ids = sdp_record_creation_pending_ids;
+    sdp_record_creation_pending_ids.clear();
+  }
+
+  for (int id : pending_ids) {
+    remove_sdp_record(id);
+  }
+}
+
 /******************************************************************************
  * CALLBACK FUNCTIONS
  * Called in BTA context to create/remove SDP records.
@@ -413,6 +440,19 @@ void on_create_record_event(int id) {
          */
         btif_enable_service(service_id);
       }
+
+      bool has_pending = false;
+
+      {
+        std::unique_lock<std::recursive_mutex> lock(sdp_lock);
+        sdp_slots[id].state = SDP_RECORD_ALLOCED;
+        has_pending = !sdp_record_creation_pending_ids.empty();
+      }
+
+      /* Check if any records were pending removal */
+      if (has_pending) {
+        sdp_check_pending_records();
+      }
     }
   }
 }
-- 
2.55.0.897.gb25b4bd76c-goog

